## What You're Aiming For

In this checkpoint, we are going to work on 'Education Attainment And Enrolment Around The World' dataset, provided by the World bank group.

Dataset description : The dataset was derived from the original copy, which was published on the world bank website. The primary purpose of this database is to document and analyse patterns of educational attainment rates in different country around the world.

➡️ Dataset link

https://i.imgur.com/bMQy55e.jpg

Columns description : 

Age[15-49]All_gradeX : Percentage of the entire population ages 15 to 49 that has completed grade X

Age[15-49]Male_gradeX : Percentage of the male population ages 15 to 49 that has completed grade X

Age[15-49]Female_gradeX : Percentage of the female population ages 15 to 49 that has completed grade X

X = [1...3]

### Instructions

Step 1: Data exploration with Pandas
Download the provided dataset
Use pandas to read the file into a pandas Dataframe.
Display general information about the dataset.
Run descriptive analysis on the dataset using the 'describe' function.

Step 2: Data exploration with ydata-profiling
Use ydata-profiling to generate a report of the provided dataset.
Look for missing values in the dataset. You can use the report generated by ydata to identify missing values.
Examine the correlations between different columns in the dataset. You can use the report generated by ydata to identify correlations.
Identify any other interesting patterns or insights in the dataset that you can discover through the ydata-profiling report.

Step 3: Summary
Write a summary of your findings based on the ydata-profiling report. Be sure to include any interesting patterns or insights you discovered in the dataset.

In [4]:
import pandas as pd
from ydata_profiling import ProfileReport


# Step 1: Data exploration with Pandas
# Read the dataset into a DataFrame
file_path = r"C:\Users\HARDEY\Documents\GOMYCODE\Python\EDUCATION_ATTAINMENT.csv"
education_df = pd.read_csv(file_path)

In [5]:
education_df

Ages[15-49]_All_grade1  Ages[15-49]_All_grade2  Ages[15-49]_All_grade3  \
0                  0.669091                0.667214                0.651812   
1                  0.485879                0.480171                0.465510   
2                  0.516431                0.514233                0.502269   
3                  0.914294                0.903528                0.871690   
4                  0.876990                0.841497                0.762864   
..                      ...                     ...                     ...   
621                0.987620                0.985260                0.978837   
622                0.994478                0.990348                0.986915   
623                0.993604                0.989401                0.983428   
624                0.994425                0.992300                0.987499   
625                0.993671                0.991859                0.985111   

     Ages[20-29]_All_grade1  Ages[20-29]_All_grade2  Ages[20-29]_All_grade3  \
0                  0.460899                0.458749                0.446925   
1                  0.267236                0.265735                0.257999   
2                  0.327775                0.326250                0.318003   
3                  0.853770                0.840792                0.810794   
4                  0.881265                0.854158                0.795892   
..                      ...                     ...                     ...   
621                0.988720                0.986780                0.981525   
622                0.991449                0.989851                0.984804   
623                0.987763                0.984321                0.977343   
624                0.991362                0.989737                0.985291   
625                0.993362                0.988796                0.983957   

     Ages[30-39]_All_grade1  Ages[30-39]_All_grade2  Ages[30-39]_All_grade3  \
0                  0.292791                0.290929                0.279100   
1                  0.191279                0.188839                0.182693   
2                  0.221981                0.220472                0.214040   
3                  0.775725                0.757109                0.713569   
4                  0.883625                0.853812                0.805231   
..                      ...                     ...                     ...   
621                0.977844                0.974075                0.967946   
622                0.987105                0.985023                0.979358   
623                0.983226                0.980478                0.973111   
624                0.984826                0.982728                0.975637   
625                0.990576                0.986393                0.976674   

     Ages[40-49]_All_grade1  ...  Ages[20-29]_Female_grade2  \
0                  0.250981  ...                   0.279915   
1                  0.187151  ...                   0.122435   
2                  0.205611  ...                   0.157744   
3                  0.786468  ...                   0.786506   
4                  0.747516  ...                   0.793379   
..                      ...  ...                        ...   
621                0.932959  ...                   0.985412   
622                0.970577  ...                   0.988470   
623                0.911083  ...                   0.981323   
624                0.967994  ...                   0.989167   
625                0.981112  ...                   0.989465   

     Ages[20-29]_Female_grade3  Ages[30-39]_Female_grade1  \
0                     0.272185                   0.111428   
1                     0.117916                   0.084555   
2                     0.151130                   0.086072   
3                     0.746073                   0.699617   
4                     0.712726                   0.808248   
..                         ...     

In [8]:
# Display general information about the dataset.
education_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 626 entries, 0 to 625
Data columns (total 38 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Ages[15-49]_All_grade1     626 non-null    float64
 1   Ages[15-49]_All_grade2     626 non-null    float64
 2   Ages[15-49]_All_grade3     626 non-null    float64
 3   Ages[20-29]_All_grade1     626 non-null    float64
 4   Ages[20-29]_All_grade2     626 non-null    float64
 5   Ages[20-29]_All_grade3     626 non-null    float64
 6   Ages[30-39]_All_grade1     625 non-null    float64
 7   Ages[30-39]_All_grade2     625 non-null    float64
 8   Ages[30-39]_All_grade3     625 non-null    float64
 9   Ages[40-49]_All_grade1     625 non-null    float64
 10  Ages[40-49]_All_grade2     625 non-null    float64
 11  Ages[40-49]_All_grade3     625 non-null    float64
 12  Ages[15-49]_Male_grade1    626 non-null    float64
 13  Ages[15-49]_Male_grade2    626 non-null    float64

In [10]:
# Run descriptive analysis on the dataset using the 'describe' function.
education_df.describe()

Ages[15-49]_All_grade1  Ages[15-49]_All_grade2  Ages[15-49]_All_grade3  \
count              626.000000              626.000000              626.000000   
mean                 0.873886                0.861490                0.837509   
std                  0.162624                0.166895                0.176555   
min                  0.251917                0.244071                0.233274   
25%                  0.831791                0.803131                0.757465   
50%                  0.948926                0.935167                0.909587   
75%                  0.984447                0.980112                0.971166   
max                  1.000000                1.000000                1.000000   

       Ages[20-29]_All_grade1  Ages[20-29]_All_grade2  Ages[20-29]_All_grade3  \
count              626.000000              626.000000              626.000000   
mean                 0.823849                0.812738                0.789965   
std                  0.200559                0.202587                0.207979   
min                  0.079610                0.079610                0.079610   
25%                  0.744915                0.723418                0.678869   
50%                  0.910132                0.896568                0.864849   
75%                  0.973777                0.967410                0.955625   
max                  1.000000                0.999824                0.999264   

       Ages[30-39]_All_grade1  Ages[30-39]_All_grade2  Ages[30-39]_All_grade3  \
count              625.000000              625.000000              625.000000   
mean                 0.766171                0.752519                0.724980   
std                  0.236563                0.237679                0.240681   
min                  0.098059                0.091358                0.085236   
25%                  0.639447                0.612464                0.561545   
50%                  0.853861                0.831246                0.794898   
75%                  0.959097                0.948259                0.922841   
max                  1.000000                1.000000                1.000000   

       Ages[40-49]_All_grade1  ...  Ages[20-29]_Female_grade1  \
count              625.000000  ...                 626.000000   
mean                 0.709166  ...                   0.790544   
std                  0.259998  ...                   0.236541   
min                  0.036549  ...                   0.029662   
25%                  0.527055  ...                   0.678596   
50%                  0.785356  ...                   0.900869   
75%                  0.937129  ...                   0.974514   
max                  1.000000  ...                   1.000000   

       Ages[20-29]_Female_grade2  Ages[20-29]_Female_grade3  \
count                 626.000000                 626.000000   
mean                    0.778972                   0.755644   
std                     0.238637                   0.243755   
min                     0.029662                   0.029662   
25%                     0.658977                   0.616380   
50%                     0.887206                   0.853414   
75%                     0.968962                   0.959495   
max                     1.000000                   0.999777   

       Ages[30-39]_Female_grade1  Ages[30-39]_Female_grade2  \
count                 625.000000                 625.000000   
mean                    0.719020                   0.704342   
std                     0.278133                   0.279059   
min                     0.060930                   0.058368   
25%                     0.534060                   0.508218   
50%                     0.815823                   0.794049   
75%                     0.958022                   0.946510   
max                     1.000000                   1.000000   

       Ages[30-39]_Female_grade3  Ages[40-49]_Female_grade1  \
count                 625.000000

## Step 2: Data exploration with ydata-profiling

In [35]:


# Check the number of rows in the dataset
num_rows = education_df.shape[0]

# Determine the sample size
sample_size = min(100, num_rows)  # Sample size should not exceed the number of rows

# Take a random sample of the determined size
sample_df = education_df.sample(n=sample_size, random_state=42)

In [37]:
sample_df

Ages[15-49]_All_grade1  Ages[15-49]_All_grade2  Ages[15-49]_All_grade3  \
448                0.983327                0.965076                0.930597   
486                0.964466                0.954242                0.929592   
215                0.993425                0.991771                0.991218   
212                0.993131                0.988257                0.987149   
481                0.793037                0.785024                0.773072   
..                      ...                     ...                     ...   
218                0.971709                0.958770                0.936279   
292                0.947645                0.932737                0.910583   
274                0.997370                0.997370                0.997127   
403                0.679303                0.667520                0.638320   
271                0.951102                0.943134                0.918419   

     Ages[20-29]_All_grade1  Ages[20-29]_All_grade2  Ages[20-29]_All_grade3  \
448                0.977597                0.966615                0.936120   
486                0.937375                0.920478                0.880161   
215                0.990948                0.986460                0.977194   
212                0.996814                0.995714                0.992866   
481                0.507068                0.503184                0.491432   
..                      ...                     ...                     ...   
218                0.949954                0.930133                0.893891   
292                0.898378                0.880183                0.848089   
274                0.995954                0.995748                0.995004   
403                0.514862                0.508029                0.488896   
271                0.945704                0.939779                0.929782   

     Ages[30-39]_All_grade1  Ages[30-39]_All_grade2  Ages[30-39]_All_grade3  \
448                0.968351                0.948259                0.902719   
486                0.889643                0.861801                0.814041   
215                0.979668                0.964152                0.944190   
212                0.996485                0.990711                0.988912   
481                0.338140                0.334153                0.324675   
..                      ...                     ...                     ...   
218                0.913567                0.879521                0.820635   
292                0.839737                0.814767                0.764180   
274                0.997861                0.997861                0.997535   
403                0.344972                0.335196                0.319367   
271                0.894505                0.886202                0.873861   

     Ages[40-49]_All_grade1  ...  Ages[20-29]_Female_grade2  \
448                0.950927  ...                   0.963542   
486                0.813812  ...                   0.919276   
215                0.979405  ...                   0.984929   
212                0.998767  ...                   0.995142   
481                0.331912  ...                   0.395776   
..                      ...  ...                        ...   
218                0.879382  ...                   0.942734   
292                0.821082  ...                   0.842455   
274                0.998596  ...                   0.998134   
403                0.256892  ...                   0.328988   
271                0.773402  ...                   0.926312   

     Ages[20-29]_Female_grade3  Ages[30-39]_Female_grade1  \
448                   0.931473                   0.957353   
486                   0.876896                   0.871400   
215                   0.976412                   0.969290   
212                   0.992605                   0.998318   
481                   0.383218                   0.249353   
..                         ...     

In [41]:
# Generate the profile report
# before generating a profile report, we are working with a sample to represent the whole. 

profile = ProfileReport(sample_df, title='Pandas Profiling Report')

In [ ]:
# Generate an HTML report
profile.to_file("sample_df.html")

#### Look for missing values in the dataset. You can use the report generated by ydata to identify missing values.

#### Examine the correlations between different columns in the dataset. You can use the report generated by ydata to identify correlations.


In [50]:
missing_values = sample_df.isnull().sum()
print("Missing values per column using isnull() and sum():")
print(missing_values)

Missing values per column using isnull() and sum():
Ages[15-49]_All_grade1       0
Ages[15-49]_All_grade2       0
Ages[15-49]_All_grade3       0
Ages[20-29]_All_grade1       0
Ages[20-29]_All_grade2       0
Ages[20-29]_All_grade3       0
Ages[30-39]_All_grade1       0
Ages[30-39]_All_grade2       0
Ages[30-39]_All_grade3       0
Ages[40-49]_All_grade1       0
Ages[40-49]_All_grade2       0
Ages[40-49]_All_grade3       0
Ages[15-49]_Male_grade1      0
Ages[15-49]_Male_grade2      0
Ages[15-49]_Male_grade3      0
Ages[20-29]_Male_grade1      0
Ages[20-29]_Male_grade2      0
Ages[20-29]_Male_grade3      0
Ages[30-39]_Male_grade1      0
Ages[30-39]_Male_grade2      0
Ages[30-39]_Male_grade3      0
Ages[40-49]_Male_grade1      0
Ages[40-49]_Male_grade2      0
Ages[40-49]_Male_grade3      0
Ages[15-49]_Female_grade1    0
Ages[15-49]_Female_grade2    0
Ages[15-49]_Female_grade3    0
Ages[20-29]_Female_grade1    0
Ages[20-29]_Female_grade2    0
Ages[20-29]_Female_grade3    0
Ages[30-39]_Female

#### Examine the correlations between different columns in the dataset. You can use the report generated by ydata to identify correlations.

The correlation between different columns in the dataset is actually a positive strong relationship. 

#### Identify any other interesting patterns or insights in the dataset that you can discover through the ydata-profiling report.

The dataset consists of various age groups and grades, specifically focusing on individuals aged 15-49, segmented further by all individuals, males, and females. For each subgroup, the performance across all grades is analyzed which also deals with age.

There are strong correlations between the scores of different grades, indicating that individuals who perform well in one grade are likely to perform well in others.

High Scores Across Grades: The means for all grades are relatively high (around 0.85 to 0.87), indicating generally high scores for this age group.

Left Skewness: All three grades show negative skewness, meaning most scores are concentrated towards the higher end, with fewer low scores. This suggests that the majority of the population scores well in all grades.

Small but Increasing Spread: The IQR and standard deviation increase slightly from grade 1 to grade 3, suggesting a broader spread of scores in higher grades.

Correlation: High correlation indicates that scores in one grade are likely predictive of scores in another grade. This consistency suggests that individuals tend to perform similarly across different grades.